### ISSUE TITLE
Model hyperparameters not passed in during `nlp.begin_training`

### ISSUE THREAD

**Issue by tokestermw** [explosion/spaCy#1915]
When training the built-in spaCy models, I use the [language object and `begin_training()`](https://github.com/explosion/spaCy/blob/aad26965bd0c5741c2b1028a6b5f724f753c7704/spacy/language.py#L424) similar to this [`train_parser` example](https://github.com/explosion/spaCy/blob/aad26965bd0c5741c2b1028a6b5f724f753c7704/examples/training/train_parser.py#L60).

In `begin_training`, I should be able to pass in a configuration dictionary and set the hyperparams.

```
cfg = {
    hidden_depth: 2,  # this should error out since only depth of 1 is allowed
}
...
    with nlp.disable_pipes(*other_pipes):  # only train parser
        optimizer = nlp.begin_training(**cfg)
```

But the `cfg` is not passed in to the components of the pipeline ([relevant code](https://github.com/explosion/spaCy/blob/aad26965bd0c5741c2b1028a6b5f724f753c7704/spacy/language.py#L454)).

Using the equivalent environmental variable works.

```
(aes) ➜  training git:(master) SPACY_PARSER_HIDDEN_DEPTH=2 python train_parser.py
Created blank 'en' model
ping
Traceback (most recent call last):
  File "train_parser.py", line 90, in <module>
    plac.call(main)
  File "/Users/apewu/writelab/aes/lib/python3.6/site-packages/plac_core.py", line 328, in call
    cmd, result = parser.consume(arglist)
  File "/Users/apewu/writelab/aes/lib/python3.6/site-packages/plac_core.py", line 207, in consume
    return cmd, self.func(*(args + varargs + extraopts), **kwargs)
  File "train_parser.py", line 61, in main
    optimizer = nlp.begin_training()
  File "/Users/apewu/writelab/aes/lib/python3.6/site-packages/spacy/language.py", line 458, in begin_training
    **cfg)
  File "nn_parser.pyx", line 835, in spacy.syntax.nn_parser.Parser.begin_training
  File "nn_parser.pyx", line 245, in spacy.syntax.nn_parser.Parser.Model
ValueError: Currently parser depth is hard-coded to 1.
```

The fix is to add the cfg to the component.

```
                proc.begin_training(get_gold_tuples(),
                                    pipeline=self.pipeline,
                                    sgd=self._optimizer,
                                    **cfg)
```

I can raise a PR if this is good. Thanks!

Related issue (how to pass hyperparams): https://github.com/explosion/spaCy/issues/1848
## Your Environment
* **spaCy version:** 2.0.5
* **Platform:** Darwin-17.3.0-x86_64-i386-64bit
* **Python version:** 3.6.3
* **Models:** en_core_web_sm

**Comment by tokestermw** [comment]
Added a PR (https://github.com/explosion/spaCy/pull/1916)

**Comment by tokestermw** [comment]
Actually, I just realized this will cause a regression if we use any model that's not part of `Parser`. :(

The Parser model can pass in `**cfg` to `begin_training` but other classifiers like `tagger` and `textcat` do not. I'm guessing because the parser has beam search config that can be configured during inference time.

I think it would be easiest to add `**cfg` to `begin_training()` to the other models.

I've added these changes to PR #1916:
https://github.com/tokestermw/spaCy/commits/bug/fix-not-passing-in-model-cfg-in-nlp
